File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
from collections import deque

def solution(s):
    # 스택
    stack = deque([])
    for ss in s:
        if ss == '(':
            stack.append(ss)
        else:
            if len(stack) == 0:
                return False
            else:
                stack.pop()
    
    # 결과
    if len(stack) > 0:
        return False
    else:
        return True

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (c == '(') stack.push('(');
            else {
                if (stack.isEmpty()) return false;   // 짝 없는 ')' 
                stack.pop();
            }
        }
        return stack.isEmpty();   // 닫히지 않은 '(' 검사
    }
}